Start working on CLI2 update for NEU (#41)
tekknolagi/isdt: assignments/02-cli-constructive.md

# Homework 2: CLI, Constructive

> This is the first of four "constructive" assignments---assigments where we ask
> you to write a program. The point of such assignments is to get you more
> comfortable building things in a safe environment so that later, in a more
> "real" setting, you believe in yourself. While we are somewhat biased, we also
> think they can even be thought of as fun.

Reviewer: ```suggestion
This is the first of four "constructive" assignments---most of which ask
you to write a small amount of code. While writing that code, you will likely
make mistakes that highlight gaps in your understanding and require you to
fill those gaps by referencing lecture notes or documentation. Not only is that
expected, it's the main reason we created these assignments. The programs
you produce are secondary to the experience you gain while producing them.

In the past couple of years, it has gotten dramatically easier to bypass the
trial-and-error portion of a programming assignment by generating a program
using a large language model like ChatGPT or Copilot. Although such tools
have their place, they do not in general help you build understanding, which is
the primary purpose of these assignments.
```

# Homework 2: CLI, Constructive

## `build.sh`: making trouble

A lot of systems programming is done in languages such as C, C++, and Rust.
These languages all have one thing in common: they are predominantly compiled
before being executed[^interpreters]. This two-step tango means that every time
you modify your program, you have to compile it before running it.

[^interpreters]: We say *predominantly* because the evaluation strategy is not
    necessarily part of the language! The language is the abstract concept and
    the implementation makes it go zoom. There are C and Rust interpreters and
    there are (for example) Ruby and Python compilers.

> That's not fun, especially if it requires some kind of arcane build flags that

Reviewer: ```suggestion
That's not fun, especially if it requires arcane compiler flags that
```

are hard to remember. Or maybe it has a lot of files and your build command is
getting unwieldy. Either way, you are going to solve your own problems today by

> *writing a build system*.

Reviewer: ```suggestion
*writing a build script*.
```

less scary?

> You need not have written C code ever before starting this part of the
> assignment. We will provide sample code to you so that you can test your build

Reviewer: ```suggestion
This part of the assignment uses the C compiler as an example, but it doesn't
require you to write C: we provide sample C code you can use to test your build
```

system. Later, though, you *will* write some C code and integrate it into your
build system.

Your job is to write a program in Bash, `build.sh`, that compiles some C
program. The minimal functional (but not acceptable to submit) solution looks
something like this:

```sh
#!/bin/sh
cc foo.c -o foo
```

If only it were that simple for your assignment. We'll get more into all this

> later in the term, but this program is not ideal: it builds `foo.c` into `foo`

Reviewer: I might try to reword this, but I'd need to think about it more. Will come back to this comment later.

every time `build.sh` is executed, even if `foo.c` has not changed. Your
program must instead only rebuild `foo` if any of its dependencies have
changed. In addition, your program must build intermediate object files for
each C file.

This means example runs might look like this:

```console?prompt=$
$ ls
build.sh  foo.c  foo.h  rng.c  rng.h
$ ./build.sh 
cc -c foo.c
cc -c rng.c
cc -o foo foo.o rng.o
$ ./build.sh 
$ touch rng.c
$ ./build.sh 
cc -c rng.c
cc -o foo foo.o rng.o
$
```

Notice two things:

* We can run `build.sh` and if nothing needs to get rebuilt, nothing will be
  rebuilt
* We can update individual C files without recompiling other C files

...

### Requirements

> Note: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
> "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
> document are to be interpreted as described in [RFC

> > 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Reviewer: Needed? I'm pretty sure students know what "must" means when it occurs in an assignment, and that's the only one you use.

## `build.sh`: making trouble

A lot of systems programming is done in languages such as C, C++, and Rust.
These languages all have one thing in common: programs written in them are
predominantly compiled before being executed[^interpreters]. This two-step
tango means that every time you modify your program, you have to compile it
before running it.

[^interpreters]: We say *predominantly* because the evaluation strategy is not
    necessarily part of the language! The language is the abstract concept and
    the implementation makes it go zoom. There are C and Rust interpreters and
    there are (for example) Ruby and Python compilers.

That's not fun, especially if it requires arcane compiler flags that are hard
to remember. Or maybe your program has a lot of files and your build command is
getting unwieldy. Either way, you are going to solve your own problems today by
*writing a build script*.

This part of the assignment uses the C compiler as an example, but it doesn't
require you to write C: we provide sample C code you can use to test your build
system. Later, though, you *will* write some C code and integrate it into your
build system.

> Your job is to write a program in Bash or Sh, `build.sh`, that compiles some C

Reviewer: I don't think we use the term "Sh" in notes or elsewhere.

program. The minimal functional (but not acceptable to submit) solution looks
something like this:

```sh
#!/bin/sh
cc -o foo foo.c rng.c
```

If only it were that simple for your assignment. We'll get more into all this
later in the term, but this program is not ideal: it builds `foo.c` and `rng.c`
into `foo` every time `build.sh` is executed, even if `foo.c` and `rng.c` have

> not changed. Your program must instead only rebuild `foo` if any of its

Reviewer: ```suggestion
not changed. The script you submit, in contrast, must only rebuild `foo` if any of its
```

$ ./build.sh 
cc -c rng.c
cc -o foo foo.o rng.o
$ ./foo 
hello, world! your randomly chosen number is 4
$
```

Notice three things:

* We can run `build.sh` and if nothing needs to get rebuilt, nothing will be
  rebuilt
* We can update individual C files without recompiling other C files (we do
  need to link them at the end, though)
* By "changed", we mean that the file's modification time is newer than its
  target; the file contents may not have changed

This is a pretty standard set of features for build systems. Many people have
put many years into making build systems great. Because we are dealing with
small code and don't have a lot of strange requirements, you are only going to

> peek a little bit into that yawning abyss. We hope that after writing your

Reviewer: I'd cut "We hope …" and everything after. I think referencing Make will be lost on them at this point in the course.

Author: :(

build script, you look at Make with a little less apprehension. You too could
write such a tool given enough time.

Unlike other build systems, you are going to put the "should we recompile this"
logic in the same file as the project-specific components. Together, they will
form the build script for `foo`.

### Requirements

This all might sound a little bit scary but we promise that explicitly laying
out the behavior like this makes for easier implementation. You can think of it
like a checklist.

> Note: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
> "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
> document are to be interpreted as described in [RFC
> 2119](https://datatracker.ietf.org/doc/html/rfc2119).

You MUST:

* Be able to build `foo` by running `./build.sh`

> * Only recompile a target if any of its dependencies' modified-time (mtime) is

Reviewer: I think this requirement (rebuilding based on modification time) warrants a paragraph in the text above. We are not yet at BLD and so the fact that build systems work like this won't be evident to many students.

Reviewer: Also maybe rephrase as "if at least one of its dependencies has been modified more recently than it"

Author: Added note about mtime

  later than the target's mtime
* Write your program so that it runs under `sh` or `bash`
* Write your program entirely in one file
* Include a shebang line at the beginning of your script for either `sh` or
  `bash`
* Use the error-handling practices we encouraged in lecture

You MUST NOT:

> * Shell out to `make` or `watchman` or `entr` or any other software that would

Reviewer: s/Shell out/Invoke/? A little less jargon-y.

Author: use

  do substantive work for you
* Modify any of the C source files we provide to you

You MAY:

* Print the commands executed by your script as they are executed
* Take an optional argument to `build.sh` to specify what target to build
* Use (small) standard Unix utilities. When in doubt, ask on Piazza

> * Define functions to make your code more readable

Reviewer: Should this one be a SHOULD?

* Support modifying the build by setting `CC`, `CFLAGS`, `LDFLAGS`, and other
  assorted environment variables
* Write data to temporary files as part of your build script. If you do, you
  MUST:
  * Use the `mktemp` utility to create them

>   * Automatically clean up the temporary files when the build script finishes

Reviewer: How strictly will you grade this in the rubric? If some command in their script fails unexpectedly and `set -e` bails, do the files still need to be deleted? What about if you Ctrl-C the script?

Basically asking if they need to know how to use `trap` to meet this requirement or not.

In general, you MAY go above and beyond if you want to, as long as your

> extended program still meets the requirements.

Reviewer: Should clarify if this will result in extra credit.

running on a POSIX system. Nearly all C library functions have their own man
pages; don't be afraid to use them!

Your program, `myls.c`, MUST take as its only argument a path to a directory
and print the name of each file[^what-is-a-file] inside that directory, one per
line. You MAY include files starting with `.` at your discretion. If your
program is run without a directory name or the given name is not a directory,
your program MUST print an error and return a nonzero exit code.

[^what-is-a-file]: This assignment asks you to print a list of files. The word
    "file" is often used to refer specifically to a *regular file*, which is
    the "normal" kind of file that holds data and shows up in `ls` with a type
    of `-`. However, the strict POSIX definition of "file" encompasses not only
    regular files but also directories, symlinks, devices, and every other
    thing that can go inside a directory. For this assignment, we are referring
    to the POSIX definition whenever we say "file." Your implementation of
    `myls.c` should consider directories, symlinks, and all other types of file
    when producing their output.

> Now, onto implementation notes. We suggest looking into a couple of different

Reviewer: ```suggestion
Now, on to implementation notes. We suggest looking into a couple of different
```

program list the contents of the current working directory. This is not
required for full marks.

This is a small program. You may feel tempted to copy it off the internet or
classmate. **Don't.** You are robbing yourself of learning. We have given you
enough function names to get started. Check out their manual pages.

## Using `build.sh` to compile `myls.c`

> Now integrate by adding `myls` as a target to `build.sh`! You MUST be able to

Reviewer: Should this be an official "Part 3"? Seems easy to miss otherwise.
